### ISSUE TITLE
Currently being generated cell output disappears once you close/reload the page

### ISSUE THREAD

**Issue by deKeijzer** [jupyterlab/jupyterlab#6003]
Maybe not so much a bug, more a quality of life request.
Would it be possible to 'refresh' the cell output, after closing/reloading the webpage?

Currently while running e.g. 

```
import time

for i in range(100):
     print(i)
     time.sleep(1)
```

When you reload the page when `i=5`, the cell output will stay at 5 and not update. However, the kernel does stay running and after this loop is done, one can just continue using the kernel of said notebook (which is convenient). 

When one is using loops to process data, it occasionally can take some hours for a cell to complete, while the output of the cell is printing statistics. It would be nice to be able to close the page, open it a could hours later and get the 'updated' or at least new cell output. 

Terminals already have this ability, which is awesome. If notebooks could get this ability, that would be great.

**Comment by jasongrout** [comment]
This is an issue with the larger Jupyter architecture and the 'right' fix requires significant changes to how Jupyter kernel interaction works. In particular, kernel messages and the notebook state should be stored in the server, not in browser. Good news, though: there is active work, notably in the nteract project, to fix this architectural issue with Jupyter.

CC @rgbkrk @captainsafia

**Comment by rgbkrk** [comment]
Yes! We're exploring this now. However, there's no firm deadline on this, as it's still formative.

We're going to lean heavily on the [implementation of sync within sagemathinc's cocalc](https://github.com/sagemathinc/cocalc/tree/master/src/smc-util/sync) which supports realtime interactive jupyter notebooks. That work is going to be extracted into a BSD-3-Clause [sync](https://github.com/sagemathinc/sync) library.

The intention is that we'll create a well documented and specified protocol that all modern jupyter frontends can use (jupyterlab, nteract, papermill, nbconvert).

Another part of this effort has been to firm up some of the kernel status and messaging with a relay with message recall built in (contrary to our current message replay mechanism): https://github.com/nteract/kernel-relay

Note: I include papermill and nbconvert as we'd ❤️ to be able to write to a notebook as part of a job in automation while seeing a live view in a separate frontend.

------- 

Additional interested participants: @willingc, @williamstein, @mpacer, @mseal

Related issues:

* https://github.com/jupyter/notebook/issues/641
* https://github.com/jupyter/notebook/issues/1647

**Comment by vidartf** [comment]
Note, that once RTC is in place, we could add a "super node" running in Node on the server to incorporate the jupyter messages into the notebook document.

**Comment by ellisonbg** [comment]
To add to what @vidartf has said, this has come up a number of times and the tentative idea is to have the kernel service take kernel messages and covert them to patches to the CRDT store, essentially acting as another "collaborator". Non stateful kernel messages (widgets, maybe others) can still be let through a propagate to the frontend.

**Comment by ellisonbg** [comment]
An important point. From my perspective, we are working with (perhaps) a different set of usage cases in designing this, which strongly affects the requirements. A couple of them:

* With a multi-activity UI, there are complex UX questions around how undo/redo works. In an initial design sessions, we identified usage cases for having both per document undo/redo, as well as global workspace undo/redo. We haven't settled the UX questions, but we wanted to build an architecture that could manage that.
* Non-stack based undo-redo. We see a UX need to be able to support undoing of arbitrary patches to enable users to work separate on different parts of an activity.
* Many activities and document types. JupyterLab is embracing a vision that notebooks are just one of many different types of documents and activities that people are using while working with code and data (visualizations, spreadsheets, etc., drawing, etc.).
* Extensibility at runtime. Because JupyterLab is extensible, we need to make it possible for frontend extensions to declare new schema for documents and activities on the fly in the real time store. Notebook and cell metadata is another great example of a place where on the fly extensibility is needed (if no collaboration is need it can be treated as an opaque JSON blob, but if collaboration is needed, it has to be included - nbgrader would be a good example).

I strongly understand the attraction of GraphQL, and we are using it in a number of new extensions for JupyterLab that have different requirements. But I am having a tough time squaring these requirements with a GraphQL + diffmatchpatch approach.
